src/osd/scrubber/scrub_machine.cc

```diff
@@ -767,38 +768,85 @@ ReplicaActive::~ReplicaActive()
  *
  * Process:
  * - if already reserved: clear existing reservation, then continue
- * - ask the OSD for the "reservation resource"
- * - if granted: mark it internally and notify the Primary.
- * - otherwise: just notify the requesting primary.
+ * - for async requests:
+ *   - enqueue the request with reserver;
+ *   - move to the ProcessingReservation state
+ *   - no reply is expected by the caller
+ * - for legacy requests:
+ *   - ask the OSD for the "reservation resource"
+ *   - if granted: move to ReplicaReserved and notify the Primary.
+ *   - otherwise: just notify the requesting primary.
+ *
+ * implementation note: sc::result objects cannot be copied or moved. Thus,
+ * we've resorted to returning a code indicating the next action.
  */
-void ReplicaActive::on_reserve_req(const ReplicaReserveReq& ev)
+ReplicaReactCode ReplicaActive::on_reserve_request(
+    const ReplicaReserveReq& ev,
+    bool async_request)
 {
   DECLARE_LOCALS;  // 'scrbr' & 'pg_id' aliases
-  dout(10) << "ReplicaActive::on_reserve_req()" << dendl;
+  dout(10) << fmt::format(
+		  "ReplicaActive::on_reserve_request(): request:{} is async?{}",
+		  ev, async_request)
+	   << dendl;
+  auto& svc = m_osds->get_scrub_services();  // shorthand
 
   if (reserved_by_my_primary) {
-    dout(10) << "ReplicaActive::on_reserve_req(): already reserved" << dendl;
-    // clear the existing reservation
-    clear_reservation_by_remote_primary();  // clears the flag, too
+    dout(10) << "ReplicaActive::on_reserve_request(): already reserved"
+	     << dendl;
+    // clear the existing reservation. Clears the flag, too
+    clear_reservation_by_remote_primary();
   }
 
-  // ask the OSD for the reservation
-  const auto ret = get_remote_reservation();
-  if (ret.granted) {
-    reserved_by_my_primary = true;
-    dout(10) << fmt::format("{}: reserved? yes", __func__) << dendl;
-  } else {
-    dout(10) << fmt::format("{}: reserved? no ({})", __func__, ret.error_msg)
+  Message* reply{nullptr};
+  ReplicaReactCode next_action{ReplicaReactCode::discard};
+
+  // should we deny the request if recovery is in progress?
+  /// \todo reconsider
+  if (false &&
+      !scrbr->get_pg_cct()->_conf.get_val<bool>("osd_scrub_during_recovery") &&
+      m_osds->is_recovery_active()) {
```

Comment:
Let's rebase this on the PR that removes that if.

Reply:
Had to remove the code directly, as the code changed too much